gears/src/types/pagination/request.rs

```diff
@@ -1,64 +1,82 @@
 use serde::Deserialize;
+use vec1::Vec1;
 
-use crate::ext::Pagination;
+use crate::ext::{Pagination, PaginationByKey, PaginationByOffset};
 
 pub(crate) const QUERY_DEFAULT_LIMIT: u8 = 100;
 
-//#[derive(FromForm, Debug)]
 #[derive(Deserialize, serde::Serialize, Debug, Clone, Eq, PartialEq)]
-pub struct PaginationRequest {
+#[serde(untagged)]
+pub enum PaginationKind {
+    /// key is a value returned in PageResponse.next_key to begin
+    /// querying the next page most efficiently. Only one of offset or key
+    /// should be set.
+    Key { key: Vec1<u8> },
     /// offset is a numeric offset that can be used when key is unavailable.
     /// It is less efficient than using key. Only one of offset or key should
     /// be set.
-    pub offset: u32,
+    Offset { offset: u32 },
+}
+
+//#[derive(FromForm, Debug)]
+#[derive(Deserialize, serde::Serialize, Debug, Clone, Eq, PartialEq)]
+pub struct PaginationRequest {
+    pub kind: PaginationKind,
     /// limit is the total number of results to be returned in the result page.
     /// If left empty it will default to a value to be set by each app.
     pub limit: u8,
 }
 
 impl From<PaginationRequest> for Pagination {
-    fn from(PaginationRequest { offset, limit }: PaginationRequest) -> Self {
-        Self {
-            offset: offset
-                .checked_mul(limit as u32)
-                .map(|this| this as usize)
-                .unwrap_or(usize::MAX),
-            limit: limit as usize,
+    fn from(PaginationRequest { kind, limit }: PaginationRequest) -> Self {
+        match kind {
+            PaginationKind::Key { key } => Self::from(PaginationByKey {
+                key,
+                limit: limit as usize,
+            }),
+            PaginationKind::Offset { offset } => Self::from(PaginationByOffset {
+                offset: offset
+                    .checked_mul(limit as u32)
+                    .map(|this| this as usize)
+                    .unwrap_or(usize::MAX),
+                limit: limit as usize,
+            }),
         }
     }
 }
 
-impl Default for PaginationRequest {
-    fn default() -> Self {
-        Self {
-            offset: 0,
-            limit: QUERY_DEFAULT_LIMIT,
-        }
-    }
-}
-
-
 impl From<core_types::query::request::PageRequest> for PaginationRequest {
     fn from(
         core_types::query::request::PageRequest {
-            key: _,
+            key,
             offset,
             limit,
             count_total: _,
             reverse: _,
         }: core_types::query::request::PageRequest,
     ) -> Self {
         Self {
-            offset: offset.try_into().unwrap_or(u32::MAX),
+            kind: match key.is_empty() {
+                true => PaginationKind::Key {
+                    key: key.try_into().expect("we checked it for empty"),
```

Comment:
Are true/false the wrong way round here?

Reply:
fixed